src/libsystemd/sd-event/test-event.c

```diff
@@ -985,4 +985,39 @@ TEST(defer_add_post) {
         ASSERT_TRUE(dispatched_post);
 }
 
+static int child_handler_wnowait(sd_event_source *s, const siginfo_t *si, void *userdata) {
+        ASSERT_OK(sd_event_exit(sd_event_source_get_event(s), 0));
+        return 0;
+}
+
+TEST(child_wnowait) {
+        _cleanup_(sd_event_unrefp) sd_event *e = NULL;
+        pid_t pid;
+        siginfo_t si;
+
+        ASSERT_OK(sigprocmask_many(SIG_BLOCK, NULL, SIGCHLD));
+
+        ASSERT_OK(sd_event_default(&e));
+
+        /* Fork a subprocess */
+        ASSERT_OK_ERRNO(pid = fork());
+
+        if (pid == 0)
+                /* Child process - exit with a specific code */
+                _exit(42);
+
+        /* Add a child source with WNOWAIT flag */
+        ASSERT_OK(sd_event_add_child(e, NULL, pid, WEXITED|WNOWAIT, child_handler_wnowait, NULL));
+
+        /* Run the event loop - this should call the handler */
+        ASSERT_OK(sd_event_loop(e));
```

Comment:
can you extend the test slightly and ensure that the handler gets called again and again? (i.e. maintain a counter, just exit after the 5th iteration)

Reply:
Even after the obvious fix of making sure the event source is set to SD_EVENT_ON, the handler still only gets called once. Are you sure you're assuming correctly what the kernel does with pidfd's here?

Reply:
hmm, this smells like a bug somewhere, no? i mean, conceptually, as long as the thing isn't reaped it really should reamain pending.

Reply:
Indeed it turned out to be a bug where we shortcutted event_source_offline() and thus the `EPOLLONESHOT` flag did not get removed. Fixed that now and updated the test.